### PR TITLE
[lexer] keyword protected quotation token for arbitrary text

### PR DESCRIPTION
**Pull request by gares**
This PR implements a feature I could use in elpi. Today the only token that can carry arbitrary text is STRING that suffers from escaping shortcomings. This PR implements a class of tokens that can carry arbitrary text, so that 3rd party languages can be easily embedded in .v files.

From the last commit:
One can now register a quotation using a grammar rule with
`QUOTATION "name:"`. In that case `name:` becomes a keyword and the token is
generated for `name:` followed by a an identifier or a parenthesized
text. Eg
```
  constr:x
  string:[....]
  ltac:(....)
  ltac:{....}
```

The delimiter is made of 1 or more occurrences of the same parenthesis,
eg `((.....))` or `[[[[....]]]]`. The idea being that if the text happens to
contain the closing delimiter, one can make the delimiter longer and avoid
confusion (no escaping). Eg

```
string:[[ .. ']' .. ]]
```

Nesting the delimiter is allowed, eg `((..((...))..))` is OK.

The text inside the quotation is returned as a string (including the
parentheses), so that a third party parser can take care of it.

Keywords don't need to end in `:`.

### CODE REVIEW

**Comment by ppedrot** [comment]
Note that this is probably going to conflict with #8764, it might be worthwhile to factor out the common parts.

**Comment by gares** [comment]
I'm fine taking the conflicts, but  #8764 seems to have stalled, I'll ping.

**Comment by gares** [comment]
Wrt the changes, they are quite orthogonal as far as I can tell.

**Comment by proux01** [comment]
> Wrt the changes, they are quite orthogonal as far as I can tell.

Indeed, but Pierre Marie is right in that some conflicts may arise (c.f. https://github.com/coq/coq/pull/8764/commits/4bc83c8f6fd3a375c5464e44d16ac39cb8ef0526).

**Comment by gares** [comment]
Sure. I was just offering to give you precedence. My change is smaller and it is hence easier to rebase. Of course this works for me only if you push your PR ahead, hence my ping.

**Comment by gares** [comment]
OK, I've tested it in elpi and it works for me. I've added the `needs:documentation` because I don't know where to put it (the commit message/PR header contains quite some doc already).

**Comment by gares** [comment]
@ppedrot BTW some of this code and/or the delimiting convention could be used in coqpp too. I had to write `... -> { ... '\x7b' ... }` somewhere in elpi in a grammar extend, since `'}'` would confuse coqpp by terminating the ML block too early. The convention proposed here would have let me just write `... -> {{ ... '}' ... }}`.

**Comment by gares** [comment]
ping @coq/parsing-maintainers  (and @coq/doc-maintainers to tell me where I should put the doc of the QUOTATION token, especially the non-escaping rules).

**Comment by Zimmi48** [comment]
This should probably go in the “Syntax extension” chapter (https://coq.inria.fr/refman/user-extensions/syntax-extensions.html) *if* it impacts the users directly. If it only concerns the plugin writers, then it should go in the non-existing plugin writing documentation chapter :smiling_imp:. Until then, you can just create a new Markdown file in `dev/doc`.

**Comment by Zimmi48** [comment]
The index in https://github.com/coq/coq/tree/master/dev#miscellaneous-information-about-the-code-devdoc should be updated if you create a new file in `dev/doc`.

**Comment by gares** [comment]
It does not concern users, unless they use an extension language that uses the feature.
So I'll write the comment in CLexer or Tok for now.

**Comment by gares** [comment]
ping @coq/parsing-maintainers for the review

File: gramlib/plexing.mli
Comment (ejgallego) [inline gramlib/plexing.mli]: The doc below needs update.

File: parsing/cLexer.mli
Comment (ejgallego) [inline parsing/cLexer.mli]: Maybe it would be clearer to have two functions ? `add_keyword` and `add_quotation` ?

**Review by ejgallego — COMMENTED** [review]
I am not expert enough in this code, I let @ppedrot or @herbelin handle this.

I saw nothing wrong.

File: gramlib/plexing.mli
Comment (gares) [reply gramlib/plexing.mli]: done

File: parsing/cLexer.mli
Comment (gares) [reply parsing/cLexer.mli]: I thought about that, but in the current model a quotation is always initiated by a keyword. So the API would look like
```
val add_keyword : string -> unit
val add_keyword_for_quotation : string -> unit
```
that is pretty much what we have in the patch...

File: parsing/cLexer.mli
Comment (Zimmi48) [inline parsing/cLexer.mli]: Try to avoid the non-gender-neutral "he". You can simply replace it by "they" (plural acting as gender-neutral singular). If you don't like this, then "she" is still preferable to "he" to acknowledge that not all developers are male.

**Review by Zimmi48 — COMMENTED** [review]
Documentation is good, I only have a minor comment.

File: parsing/cLexer.mli
Comment (gares) [reply parsing/cLexer.mli]: done

**Comment by gares** [comment]
This till lacks an assignee. @ppedrot can you take it?

**Comment by gares** [comment]
ping

**Comment by gares** [comment]
ping @ppedrot (time for 8.10 is running out)

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
@gares Can you rebase? I'll merge it after.

**Comment by gares** [comment]
I did rebase the overlay, the error seems from another pr

**Comment by Zimmi48** [comment]
The error seems very related to this PR to me:

```
File "./src/Rewriter.v", line 1347, characters -3739--3739:
Warning: Not interpreting "*)" as the end of current non-terminated comment
because it occurs in a non-terminated string of the comment.
[comment-terminator-in-string,parsing]
File "./src/Rewriter.v", line 3585, characters -1686-0:
Error: Syntax Error: Lexer: Unterminated string
```

**Comment by gares** [comment]
Ah ok, the Ltac2 one is fixed now.
The one of Fiat seems indeed related, I look into it tomorrow. It is still weird that it passed CI a week ago...

**Comment by proux01** [comment]
Even weirder, I just rebased #9815 over this PR and CI passes (with just a simple overlay for ltac2).

**Comment by gares** [comment]
Restarted, and now it is green... I guess they pushed a bad commit

**Pull request merged**
